### PR TITLE
fMP metric includes new blank webfont signal

### PR DESCRIPTION
**Pull request by paulirish**
This matches up with the impl in traceviewer: https://codereview.chromium.org/1998063002
We require layouts that have fewer than 200 invisible text characters, via `approximateBlankCharacterCount`

Also we report all fMP candidate paints in the extendedInfo.

image
Here's a run with https://acrobat.adobe.com/us/en/sign.html?promoid=HM85XGHT which has some nice FOIT

### CODE REVIEW

File: src/audits/performance/first-meaningful-paint.js [hunk 1]
```diff
@@ -205,6 +211,11 @@ class FirstMeaningfulPaint extends Audit {
         return (max(1, ratioBefore) + max(1, ratioAfter)) / 2;
       }
 
+      function hasTooManyBlankCharactersToBeMeaningful() {
```
Comment (brendankenny) [inline src/audits/performance/first-meaningful-paint.js]: comment to explain what this function is doing?

File: src/audits/performance/first-meaningful-paint.js [hunk 2]
```diff
@@ -114,13 +117,15 @@ class FirstMeaningfulPaint extends Audit {
...
+    let timings = {};
+    Object.keys(data.fmpCandidates).forEach(name => {
+      const evt = data.fmpCandidates[name];
+      timings[name] = evt && ((evt.ts - data.navStart.ts) / 1000);
+    });
 
-    // First meaningful paint
-    const firstMeaningfulPaint = (lastfMPts - data.navStart.ts) / 1000;
+    // First meaningful paint is the last timestamp observed from the candidates
+    const firstMeaningfulPaint = Object.values(timings).reduce((mx, c) => max(mx, c));
```
Comment (brendankenny) [inline src/audits/performance/first-meaningful-paint.js]: `Object.values` -- This appears to still be behind a flag, even in node 6/v8 5.3 if I'm not missing something

Comment (brendankenny) [inline src/audits/performance/first-meaningful-paint.js]: also, some better variable names here. `mx` is max? What is `c`?

**Comment by brendankenny** [comment]
looks good with some nits and a need to work around lack of `Object.values`

**Comment by mgerakis** [comment]
It seems that some errors with first paint are still occurring. Here is an example of the worst case, where the speed index is -1:
`Page load performance is fast: NaN%
 -- First meaningful paint: NaN (-Infinityms)
 -- Speed Index: -1
    Cannot read property 'getImage' of undefined`

**Comment by mgerakis** [comment]
I believe the issue is here: 
**src/scheduler.js**
74: `tracingData.traceContents = traceContents;`

From what I can see, when the error occurs the tracing data that contains firstContentfulPaint and Paint events are happening in different frames. However, the mainFrameID cannot change once set:

172: `if (event.name === 'navigationStart' && !navigationStart) {`
173: `    mainFrameID = event.args.frame;`
174:`     navigationStart = event;`
175:`}`

**src/audits/performance/first-meaningful-paint.js**
176: `if (event.name === 'firstContentfulPaint' && event.args.frame === mainFrameID)`
186: `if (event.name === 'Paint' && event.args.data.frame === mainFrameID)`

**Comment by mgerakis** [comment]
https://github.com/mgerakis/lighthouse/tree/first-paint-fix

^ This link has the necessary changes to handle multiple frames and correctly record the first meaningful paint.

**Comment by samccone** [comment]
@mgerakis mind opening a PR from your branch?

File: src/audits/performance/first-meaningful-paint.js [hunk 3]
```diff
@@ -27,6 +27,8 @@ const FAILURE_MESSAGE = 'Navigation and first paint timings not found.';
 const SCORING_POINT_OF_DIMINISHING_RETURNS = 1600;
 const SCORING_MEDIAN = 4000;
 
+const BLOCK_FIRST_MEANINGFUL_PAINT_IF_BLANK_CHARACTERS_MORE_THAN = 200;
```
Comment (paullewis) [inline src/audits/performance/first-meaningful-paint.js]: What's this used for?

File: src/audits/performance/first-meaningful-paint.js [hunk 4]
```diff
@@ -213,7 +230,7 @@ class FirstMeaningfulPaint extends Audit {
       // layout significance = number of layout objects added / max(1, page height / screen height)
       significance = (heuristics.pageHeight) ? (layoutCount / heightRatio()) : layoutCount;
 
-      if (heuristics.webFont && counter('hasBlankText')) {
+      if (heuristics.webFont && hasTooManyBlankCharactersToBeMeaningful()) {
```
Comment (paullewis) [inline src/audits/performance/first-meaningful-paint.js]: OIC so if you're waiting on a web font and you've got a load of blank characters you're not happy. Nice!

**Comment by paullewis** [comment]
Reviewed. Happy to +1 once @brendankenny's nits are dealt with.

**Comment by paullewis** [comment]
FWIW there's an [Object.values polyfill over at the TC39](https://github.com/tc39/proposal-object-values-entries/blob/master/polyfill.js) which we can put in third_party

**Comment by paulirish** [comment]
Comments addressed

Here's a run showing how the combination of candidates work well:

```sh
lighthouse --quiet --output=json http://hellohappy.org/beautiful-web-type/ | grep -A 3 -B 13 "first-meaningful-paint" | jq .
```

image

**Comment by paulirish** [comment]
Can I get a review? I think this is in good shape to land.

**Comment by brendankenny** [comment]
looks good!

**Pull request merged**